include/tvm/ir/module.h

```diff
@@ -56,11 +56,14 @@ class IRModuleNode : public Object {
   Map<GlobalTypeVar, TypeData> type_definitions;
   /*! \brief The source map for the module. */
   parser::SourceMap source_map;
+  /*! \brief The names of external functions for func registry */
+  Array<String> ext_func_names;
```

Comment:
why is this needed? I think external functions should be able to get from the IRModule as well.

Reply:
IIUC, by the time IRModule reaches codegen_c_host (See line 311), all the functions are PrimFunc s. Thus, it will not include the relay external functions that were taken out early in the compilation flow. However, when creating the system-lib those function names should be included in the func registry as they are also sort of "primitive" but externally compiled.

Reply:
yeah, my point was that we probably don't need to have this in Relay.

Reply:
@zhiics I think i see what's happening here: with the c++ runtime on full OS, codegen could generate symbols from a variety of modules and rely on runtime lookup via dlsym. with C runtime which has no dlsym, we use a FuncRegistry which is currently generated inside the host module. this has led to the need for "model-level" metadata being passed to the codegen in a couple of cases. probably the right thing to do is move FuncRegistry to its own module and pass this metadata there?

Reply:
ahh, I see. Then we don't really need to have changes in Relay if we move FuncRegistry to its own module and pass the functions names there, right?

Reply:
I think that is the right way to go...I don't have bandwidth for that til after TVM conf though :/

Reply:
I would suggest we wait a bit to get on the right direction. @manupa-arm how do you think?

Reply:
Would it be possible to mark this as a TODO (if that's not breaking or conflicting anything) and move on? IIUC, the suggestion is we would create another module that ends up being a source with func registry on it.

Reply:
OK then, how about we use the MetaData module wrapper to capture the function names (inc. external ones -- as it has access to all the modules) and emit to devc.cc / devc.o ?

Reply:
I think the issue with metadata module is that it assumes it is just generating the data for storage rather than immediate consumption on the target device. it requires a load process that ultimately copies the data into RAM. we want to avoid this at all cost on µTVM, because RAM is most precious. so I think we either would need to modify metadatamodule to enforce alignment or create something else.

Reply:
That is true for Packing of Imports which get unpacked in the runtime to recreate the runtime modules. I guess the question is would that be the only purpose of the metadata module ? since it is named as metadata module. Thus, here I am not suggesting not to pack, instead make the metadata module generate the func registry in the artifact (which is devc.cc/o) that it uses to pack the imports.

Reply:
So this is my understanding, metadata module will pack non-DSO exportable imported modules and their constants to devc.{object_format) if they implement "get_symbol" and "get_const_vars", otherwise its virtually doing nothing (Correct me if I am wrong, @zhiics). In this case, when we want to create external c-source modules from our external codegen, metadata module's load process should be virtually harmless -- as the SavetoBinary will almost write nothing to devc.{object_format} file (but there are things being written)

So my question is since metadata module holds all the modules (DSO and non-DSO exportable which are just llvm/c modules) can generate the func_registry as well, if we want another entity to deal with model-level metadata. Do we really need another module for this ? I would really appreciate a way forward on this.  also : @tqchen

Reply:
Sorry for the late response since it was a holiday. you are right, metadata module has the symbols and constants. BTW, are the symbols enough for you? It looks that you just need them to create the declarations.

Reply:
Hi @zhiics , no worries and hope you had a good holiday :) 

I actually went down on a path to create a CSourceMetadata module that could encapsulate c-source modules and which in-turn could get encapsulated again by the current metadata module if some of the external function require constant offloading -- I think that could be decided based on the fact whether the external module implements "get_const_vars" and that has non-zero constants. I think if dont have constants that the external module requires to offload metadata module, we dont need to create the current metadata module.

Summarizing, Im thinking of having two layers of metadata modules:

dsoModule1      dsoModule2
|                              |
|                              |
 ----         |          -----
               |
 C-Source metadata module                      non-dso module (e.g., acl)
               |                                                    (that needs constant offloading)
               |                                                                          |
                 ------------------------| ------------------------------------
                                              |
                                    Metadata module

Thus, the first layer can hold the func registry (currently its implemented in both codegen_c_host and codegen_llvm). I think it could be just live in the C-source metadata module that can get compiled along with other DSOExportable modules. **I will put and RFC with a different PR to implement this change**. This way we also get to use the init process to do certain initialization for the BYOC external device.

Reply:
Sorry :p tried my best drawing with text, but I think you should get what I mean.

Reply:
@manupa-arm Thanks for more explanations. It looks good to me. This is exactly how metadata module has been used.

Reply:
@manupa-arm that also seems reasonable to me. one question though--do you still intend to keep FuncRegistry codegen as-is (i.e. would you still generate the same global symbols)? it would be nice to avoid worrying about alignment issues for global symbols if we are generating C.

Reply:
No I've already to removed the FuncRegistry from codegen_host_c and codegen_llvm and make the runtime modules expose a packed func named : "get_func_names"  -- we can discuss the best name for it , I am not good at naming stuff :p . So the CSourceMetadata module will use from the runtime modules and create a single func registry as a seperate source to be linked together. Still fixing some tests, I ll put it up soon.

Reply:
ok, looking forward to it. sorry what I mean is: do you intend to output the same thing, even if it's just from a different module?

enumeration of the functions a TVMModule contains is a standing problem with the PackedFunc module infra. cc @tqchen @jroesch

Reply:
Sorry, I am not sure I follow. Here what I meant is we could create the function registry in a different source that captures the function names from all the runtime modules and the packed func is also used only in the compile time to create c-source metadata module. That would be exactly same as what you would've produced inside the main "c"/"llvm" module except its created in a different module that has access to all the runtime modules created.

Reply:
cool--that makes sense to me.

Reply:
https://github.com/apache/tvm/pull/7002/ -- Implemented here.

Reply:
This is merged; hence resolving the conversation